flowmachine/flowmachine/features/location/labelled_spatial_aggregate.py

```diff
@@ -0,0 +1,117 @@
+# This Source Code Form is subject to the terms of the Mozilla Public
+# License, v. 2.0. If a copy of the MPL was not distributed with this
+# file, You can obtain one at http://mozilla.org/MPL/2.0/.
+from typing import List
+
+from flowmachine.core import Query
+from flowmachine.core.mixins import GeoDataMixin
+from flowmachine.features.location.spatial_aggregate import SpatialAggregate
+
+
+class LabelledSpatialAggregate(GeoDataMixin, Query):
+    """
+    Class representing a disaggregation of a SpatialAggregate by some label or set of labels
+
+    Parameters
+    ----------
+    locations: Query
+        Any query with a subscriber and location columns
+    subscriber_labels: Query
+        Any query with a subscriber column
+    label_columns: List[str]
+        A list of columns in subscriber_labels to aggregate on
+
+    Examples
+    --------
+
+    >>> locations = locate_subscribers(
+    ...     "2016-01-01",
+    ...     "2016-01-02",
+    ...     spatial_unit=make_spatial_unit("admin", level=3),
+    ...     method="most-common",
+    ... )
+    >>> metric = SubscriberHandsetCharacteristic(
+    ...     "2016-01-01", "2016-01-02", characteristic="hnd_type"
+    ... )
+    >>> labelled = LabelledSpatialAggregate(locations=locations, subscriber_labels=metric)
+    >>> labelled.get_dataframe()
+               pcod label_value  value
+     0  524 3 08 44     Feature     36
+     1  524 3 08 44       Smart     28
+     2  524 4 12 62     Feature     44
+     3  524 4 12 62       Smart     19'
+
+    """
+
+    def __init__(
+        self,
+        locations: Query,
+        subscriber_labels: Query,
+        label_columns: List[str] = ("value",),
+    ):
+
+        if any(
+            label_column not in subscriber_labels.column_names
+            for label_column in label_columns
+        ):
+            raise ValueError(
+                f"One of {label_columns} not a column of {subscriber_labels}"
+            )
+        if "subscriber" not in locations.column_names:
+            raise ValueError(f"Locations query must have a subscriber column")
+        if not hasattr(locations, "spatial_unit"):
+            raise ValueError(f"Locations must have a spatial_unit attribute")
+        if "subscriber" in label_columns:
+            raise ValueError(f"'subscriber' cannot be a label")
+
+        self.locations = locations
+        self.subscriber_labels = subscriber_labels
+        self.label_columns = list(label_columns)
+        self.spatial_unit = locations.spatial_unit
+        self.label_columns = label_columns
+        self.out_label_columns = [
+            f"label_{label_col}" for label_col in self.label_columns
+        ]
+        super().__init__()
+
+    @property
+    def column_names(self) -> List[str]:
+        return (
+            list(self.spatial_unit.location_id_columns)
+            + self.out_label_columns
+            + ["value"]
+        )
+
+    @property
+    def out_label_columns_as_string_list(self):
```

Comment:
Is it worth having this as a property? Given that this is only used in `RedactedLabelledSpatialAggregate`, I'd be inclined to do the list-to-string bit there.

Reply:
I think it could be useful for building queries based on labels or propagating labels through queries for the future

Reply:
Yes, possibly. OK, that seems fair enough.